test/solver/solver.cpp

```diff
@@ -0,0 +1,979 @@
+                                   int stride)
+    {
+        auto size = gko::dim<2>{mtx.ref->get_size()[0],
+                                static_cast<gko::size_type>(nrhs)};
+        auto result = VecType::create(ref, size, stride);
+        result->read(gen_dense_data<typename VecType::value_type,
+                                    typename Mtx::index_type>(size));
+        return {std::move(result), exec};
+    }
+
+    template <typename VecType>
+    double tol(const test_pair<VecType>& x)
+    {
+        return Config::tolerance() * std::sqrt(x.ref->get_size()[1]);
+    }
+
+    template <typename VecType>
+    double mixed_tol(const test_pair<VecType>& x)
+    {
+        return std::max(r_mixed<value_type, mixed_value_type>() *
+                            std::sqrt(x.ref->get_size()[1]),
+                        tol(x));
+    }
+
+    template <typename TestFunction>
+    void forall_matrix_scenarios(TestFunction fn)
+    {
+        auto guarded_fn = [&](auto mtx) {
+            try {
+                fn(std::move(mtx));
+            } catch (std::exception& e) {
+                FAIL() << e.what();
+            }
+        };
+        {
+            SCOPED_TRACE("Empty matrix (0x0)");
+            guarded_fn(gen_mtx(0, 0, 0, 0));
+        }
+        {
+            SCOPED_TRACE("Sparse Matrix with variable row nnz (50x50)");
+            guarded_fn(gen_mtx(50, 50, 10, 50));
+        }
+    }
+
+    template <typename TestFunction>
+    void forall_solver_scenarios(const test_pair<Mtx>& mtx, TestFunction fn)
+    {
+        auto guarded_fn = [&](auto solver) {
+            try {
+                fn(std::move(solver));
+            } catch (std::exception& e) {
+                FAIL() << e.what();
+            }
+        };
+        {
+            SCOPED_TRACE("Defaulted solver");
+            guarded_fn(test_pair<SolverType>{Config::build(ref, 0)
+                                                 .on(ref)
+                                                 ->generate(mtx.ref)
+                                                 ->create_default(),
+                                             Config::build(exec, 0)
+                                                 .on(exec)
+                                                 ->generate(mtx.dev)
+                                                 ->create_default()});
+        }
+        {
+            SCOPED_TRACE("Cleared solver");
+            test_pair<SolverType> pair{
+                Config::build(ref, 0).on(ref)->generate(mtx.ref),
+                Config::build(exec, 0).on(exec)->generate(mtx.dev)};
+            pair.ref->clear();
+            pair.dev->clear();
+            guarded_fn(std::move(pair));
+        }
+        {
+            SCOPED_TRACE("Unpreconditioned solver with 0 iterations via clone");
+            guarded_fn(test_pair<SolverType>{
+                Config::build(ref, 0).on(ref)->generate(mtx.ref), exec});
```

Comment:
Missing device element ?

Reply:
that is being generated implicitly by the test_pair constructor calling ref->clone(exec);